src/Stratis.SmartContracts.Executor.Reflection/ILRewrite/ObserverRewriter.cs

```diff
@@ -0,0 +1,272 @@
+        private static readonly HashSet<OpCode> BranchingOps = new HashSet<OpCode>
+        {
+            OpCodes.Beq,
+            OpCodes.Beq_S,
+            OpCodes.Bge,
+            OpCodes.Bge_S,
+            OpCodes.Bge_Un,
+            OpCodes.Bge_Un_S,
+            OpCodes.Bgt,
+            OpCodes.Bgt_S,
+            OpCodes.Ble,
+            OpCodes.Ble_S,
+            OpCodes.Ble_Un,
+            OpCodes.Blt,
+            OpCodes.Bne_Un,
+            OpCodes.Bne_Un_S,
+            OpCodes.Br,
+            OpCodes.Brfalse,
+            OpCodes.Brfalse_S,
+            OpCodes.Brtrue,
+            OpCodes.Brtrue_S,
+            OpCodes.Br_S
+        };
+
+        private static readonly HashSet<OpCode> CallingOps = new HashSet<OpCode>
+        {
+            OpCodes.Call,
+            OpCodes.Calli,
+            OpCodes.Callvirt
+        };
+
+        /// <summary>
+        /// Holds the key to retrieve the last Observer instance created from <see cref="ObserverInstances"/>. 
+        /// 
+        /// TODO: There should be a better pattern here to implement the Rewriter interface and still be able to retrieve this.
+        /// </summary>
+        public Guid LastRewritten { get; private set; }
+
+        public ModuleDefinition Rewrite(ModuleDefinition module)
+        {
+            Guid id = Guid.NewGuid();
+
+            FieldDefinition observerInstanceField = GetObserverInstance(module, id);
+            var observer = new ObserverReferences(observerInstanceField, module);
+
+            foreach (TypeDefinition type in module.Types)
+            {
+                RewriteType(type, observer);
+            }
+
+            this.LastRewritten = id;
+
+            return module;
+        }
+
+        /// <summary>
+        /// Inserts a static type into the module which gives access to an instance of <see cref="Observer"/>.
+        /// Because this is injected per module, it counts as a separate type and will not be a shared static.
+        /// </summary>
+        private FieldDefinition GetObserverInstance(ModuleDefinition module, Guid id)
+        {
+            // Add new type that can't be instantiated
+            var instanceType = new TypeDefinition(
+                "<Stratis>", "<RuntimeObserverInstance>",
+                TypeAttributes.Class | TypeAttributes.Abstract | TypeAttributes.Sealed | TypeAttributes.NotPublic,
+                module.ImportReference(typeof(object))
+            );
+
+            // Add a field - an instance of our Observer!
+            var instanceField = new FieldDefinition(
+                "Instance",
+                FieldAttributes.Assembly | FieldAttributes.Static | FieldAttributes.InitOnly,
+                module.ImportReference(typeof(Observer))
+            );
+            instanceType.Fields.Add(instanceField);
+
+            // When this type is created, retrieve the Observer from our global static dictionary so it can be used.
+            var constructor = new MethodDefinition(
```

Comment:
Looks like this might replace any existing static constructors. Not a problem at the moment as we don't allow static constructors, but potentially worthy of a comment.